### PR TITLE
Fix overflow in softmax

### PR DESCRIPTION
**Pull request by starkwj**
In "_multiclass_" prediction, lleaves generates _NaN_ for some test cases.
This is caused by the current implementation of softmax calculation, which generate _Inf_ in exp: https://github.com/siboehm/lleaves/blob/a82ec5599fef67b707f701f6ad65d59890d53271/lleaves/compiler/codegen/codegen.py#L371

This PR fixes this overflow issue according to softmax in LightGBM, where each arg of exp should minus the max of args: https://github.com/microsoft/LightGBM/blob/2f60e115a899a09d91e3710fce085444d3c740a3/tests/python_package_test/utils.py#L136-L139

### CODE REVIEW

File: lleaves/compiler/codegen/codegen.py [hunk 1]
```diff
@@ -317,6 +317,7 @@ def _populate_objective_func_block(
     llvm_copysign = builder.module.declare_intrinsic(
         "llvm.copysign", (DTYPE, DTYPE), ir.FunctionType(DTYPE, (DTYPE, DTYPE))
     )
+    llvm_maxnum = builder.module.declare_intrinsic("llvm.maxnum", (DTYPE, DTYPE), ir.FunctionType(DTYPE, (DTYPE, DTYPE)))
```
[the remark this answers is not in the document]
Comment (starkwj) [reply lleaves/compiler/codegen/codegen.py:320]: Thanks, I agree with you.
However, `maximum` cannot work for me (and maybe a common case). When I try `maximum`, It turns out as "LLVM ERROR: Cannot select: t65: f64 = fmaximum t55, t64".
I tried to find out the cause, listed below. I'am not familiar with llvm, so I cannot ensure the below is correct.

1. The issue is simliar to https://github.com/llvm/llvm-project/issues/53353 . From replies, it seems that maximum of float was supported not long ago: https://github.com/llvm/llvm-project/commit/a82d27a9a6853c96f857ba0f514a78cd03bc5c35
2. For _llvmlite_, I installed the latest version 0.43.0 by pre-built binary (_pip_ or _conda_), and the required LLVM is integrated by llvmlite binary (https://llvmlite.readthedocs.io/en/latest/admin-guide/install.html). I believe this is a common usage. The integrated LLVM doesn't seems to support the float version of `maximum`.

[the remark this answers is not in the document]
Comment (siboehm) [reply lleaves/compiler/codegen/codegen.py:320]: Huh yeah it seems this is not implement. I guess it doesn't super matter, since we're currently doing this elementwise anyway (instead of e.g. using `llvm.vector.reduce.fmax`) you could just implement it via `fcmp` followed by `select`.

**Comment by siboehm** [comment]
I fixed this in #83 in a slightly simpler way. Do you want to test if it works for you before I make a new release?

**Comment by starkwj** [comment]
> I fixed this in #83 in a slightly simpler way. Do you want to test if it works for you before I make a new release? Thanks a lot for finding the bug and pointing it out!

Thanks a lot!
But I have a doubt:
https://github.com/siboehm/lleaves/blob/7b38bacaaf04c368e9013be18f285f0441921a9d/lleaves/compiler/codegen/codegen.py#L371-L375

If args[0] is not a NaN, but args[1] is NaN, `fcmp_ordered` returns false and `select` chooses args[0], so NaN is still not propagated.

However, I just noticed that the following statements of `exp` and `fsub` can still propagate NaN in args:
`result = [builder.call(llvm_exp, [builder.fsub(arg, arg_max)]) for arg in args]`.
Thus, maybe it's not necessary to handle NaN when selecting max?

**Comment by siboehm** [comment]
Yes I think you're right. I don't think there's a way to get NaNs from the max without doing an explicit NaN check. But since the denom will be NaN if any of the values is NaN this won't really matter

**Comment by siboehm** [comment]
Also this has gone out as `v1.2.3` on PyPI, conda-forge should be releasing later today
